ikarus/io/resultevaluators.hh

```diff
@@ -26,29 +26,36 @@ namespace Ikarus::ResultEvaluators {
  * @brief Struct for calculating von Mises stress
  * @ingroup resultevaluators
  * @details The VonMises struct provides a function call operator to calculate von Mises stress.
- * @remark  Only 2D stresses are supported
- * @tparam ElementType Type of the finite element
- * @tparam FERequirements Type representing the requirements for finite element calculations
- * @tparam size Size of the stress vector
- * @tparam ScalarType Scalar type for numerical calculations
+ * @tparam dim dimension of stress state
  */
+template <int dim>
+requires(dim == 2 or dim == 3)
 struct VonMises
 {
-  template <typename ElementType, typename FERequirements, int size, typename ScalarType>
-  double operator()(const ElementType& fe, const ResultRequirements<FERequirements>& req,
-                    const Dune::FieldVector<ScalarType, size>& pos, [[maybe_unused]] int comp) const
-  requires(size == 2)
-  {
-    ResultTypeMap<ScalarType> res_;
-    fe.calculateAt(req, pos, res_);
-
-    const auto& [resultType, sigma] = res_.getSingleResult();
-    assert(resultType == ResultType::cauchyStress or resultType == ResultType::PK2Stress);
-    const auto s_x  = sigma(0, 0);
-    const auto s_y  = sigma(1, 0);
-    const auto s_xy = sigma(2, 0);
-
-    return std::sqrt(std::pow(s_x, 2) + Dune::power(s_y, 2) - s_x * s_y + 3 * Dune::power(s_xy, 2));
+  /**
+   * @brief Calculate the result quantity (von Mises stress)
+   * @param resultArray EigenMatrix containing the stress state
+   * @param comp component of result (not used here)
+   * @return von Mises stress
+   */
+  double operator()(const auto& resultArray, [[maybe_unused]] const int comp) const {
+    if constexpr (dim == 2) {
+      const auto s_x  = resultArray(0, 0);
+      const auto s_y  = resultArray(1, 0);
+      const auto s_xy = resultArray(2, 0);
+
+      return std::sqrt(Dune::power(s_x, 2) + Dune::power(s_y, 2) - s_x * s_y + 3 * Dune::power(s_xy, 2));
+    } else {
+      const auto s_x  = resultArray(0, 0);
+      const auto s_y  = resultArray(1, 0);
+      const auto s_z  = resultArray(2, 0);
+      const auto s_yz = resultArray(4, 0);
```

Comment:
I suppose you didn't skip index 3 on purpose, did you? This would also be nice if it is tested somewhere.

Reply:
Can I do this in a different PR? I also want to add principal stress for 3d

Reply:
For the testing yes. But please fix the index^^